lib/windowsexec/syscall_windows.go

```diff
@@ -0,0 +1,160 @@
+	// Use to indicate that the hProcess member receives the process handle.
+	// This handle is typically used to allow an application to find out when a
+	// process created with ShellExecuteEx terminates. In some cases, such as
+	// when execution is satisfied through a DDE conversation, no handle will be
+	// returned. The calling application is responsible for closing the handle
+	// when it is no longer needed.
+	SEE_MASK_NOCLOSEPROCESS = 0x40
+)
+
+// RunAsAndWait uses `ShellExecuteExW` to create a new process with elevated
+// privileges on Windows. It then waits until that process completes, and returns
+// an error if it exited in a bad state.
+func RunAsAndWait(file string, directory string, timeout time.Duration, parameters ...string) error {
+	// Convert our various string inputs to UTF16Ptrs
+	lpVerb, err := syscall.UTF16PtrFromString("runas")
+	if err != nil {
+		return trace.Wrap(err, "converting verb to ptr")
+	}
+	lpFile, err := syscall.UTF16PtrFromString(file)
+	if err != nil {
+		return trace.Wrap(err, "converting file to ptr")
+	}
+	lpDirectory, err := syscall.UTF16PtrFromString(directory)
+	if err != nil {
+		return trace.Wrap(err, "converting directory to ptr")
+	}
+	lpParameters, err := syscall.UTF16PtrFromString(strings.Join(parameters, " "))
+	if err != nil {
+		return trace.Wrap(err, "converting parameters to ptr")
+	}
+
+	// https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shellexecuteexw
+	// Results are returned back into info :)
+	info := &shellExecuteInfoW{
+		fMask:        SEE_MASK_NOCLOSEPROCESS,
+		lpVerb:       uintptr(unsafe.Pointer(lpVerb)),
+		lpFile:       uintptr(unsafe.Pointer(lpFile)),
+		lpParameters: uintptr(unsafe.Pointer(lpParameters)),
+		lpDirectory:  uintptr(unsafe.Pointer(lpDirectory)),
+		nShow:        windows.SW_NORMAL,
+	}
+	// set the size field of info to the size of info.
+	info.cbSize = uint32(unsafe.Sizeof(*info))
+
+	success := shellExecuteExW(info)
+	if !success {
+		err := syscall.GetLastError()
+		// The above err can still be nil in certain types of failure
+		// if it is returned, then it is much more descriptive, so we should
+		// return that.
+		if err != nil {
+			return trace.Wrap(err, "calling shellExecuteExW")
+		}
+		if info.hInstApp == SE_ERR_ACCESSDENIED {
+			err := trace.BadParameter("shellExecuteExW failed with ACCESSDENIED")
+			return trace.WithUserMessage(
+				err,
+				"This error can occur if the UAC dialogue is rejected or if it is not possible to open a UAC dialogue due to system configuration.",
+			)
+		}
+		// If GetLastError is nil, the only thing we can do is push out the
+		// value of hInstApp.
+		return trace.Errorf("shellExecuteExW failed and did not call SetLastError. hInstApp=%d", info.hInstApp)
+
+	}
+	// Since we provided SEE_MASK_NOCLOSEPROCESS, closing info.hProcess is our
+	// responsibility.
+	defer syscall.CloseHandle(info.hProcess)
+
+	waitTime := syscall.INFINITE
+	if timeout != time.Duration(0) {
```

Comment:
What is this doing here?

Reply:
This was mostly to support letting callers of `RunAsAndWait` not specify a timeout. I think this is a case of YAGNI, I'll throw an error if they pass a zero value in.

Reply:
TIL what YAGNI means. Ah that makes sense, I forgot `timeout` was a parameter

Reply:
88c58ac55f